File: adb_command.py
# encoding: UTF-8
import os
import sys
import subprocess
import time
import re           #正则表达式

ads = 'adb shell '

def _command(command):
    #os.system('adb devices')
    #os.popen('adb devices').read()
    status, output = subprocess.getstatusoutput(command)
    return output

#输出设备信息
def deviceInfo():
    brand = _command(ads + 'getprop ro.product.brand')
    name = _command(ads + 'getprop ro.product.name')
    model = _command(ads + 'getprop ro.product.model')
#    print('%s %s (%s)' % (brand,model,name))
    print ('品牌:' + brand)
    print ('型号:' + model)
    print ('代号:' + name)
    print ('系统版本:' + _command(ads + 'getprop ro.build.version.release') + ',' + _command(ads + 'getprop ro.build.version.sdk'))
    print ('屏幕密度:' + _command(ads + 'wm density'))
    print ('屏幕分辨率:' + _command(ads + 'wm size'))

#输出IP地址
def deviceIP():
    os.system(ads + 'ifconfig | grep Mask')

#屏幕截图
def screencap():
    fileName = time.strftime("%Y_%m_%d_%H_%M_%S", time.localtime());
    os.system(ads + 'screencap -p /sdcard/' + fileName + '.png')
    os.system('adb pull /sdcard/' + fileName + '.png')
    os.system(ads + 'rm /sdcard/' + fileName + '.png')

#录制屏幕
def screenrecord():
    print('屏幕录制中...需要停止时按ctrl+c')
    fileName = time.strftime("%Y_%m_%d_%H_%M_%S", time.localtime());
    os.system(ads + 'screenrecord /sdcard/' + fileName + '.mp4')
    time.sleep(1)
    os.system('adb pull /sdcard/' + fileName + '.mp4')
    os.system(ads + 'rm /sdcard/' + fileName + '.mp4')

#显示包列表
#word:搜索关键字
def packages(word):
    if len(word)==0:
        os.system(ads + 'pm list packages')
    else:
        os.system(ads + 'pm list packages | grep ' + word)

def apkInfo(package):
    if len(package)==0:
        print('抱歉,请输入你要搜索的包名')
    else:
        com = ads + 'pm list packages ' + package + "| awk -F ':' '{print $2}'"
        packagesStr = _command(com)

        if len(packagesStr)==0:
            print('没有匹配到包:' + package)
        else:
            packages = packagesStr.split('\n')
            for index in range(len(packages)):
                print(index,packages[index])

            num = int(input("请选择你要查询应用:"))
            os.system("adb shell dumpsys package " + packages[num])

def inputtext(text):
    os.system("adb shell input text " + text)


def apkUtil(package):
    print('apkUtil')
    #基本信息:版本,安装时间
    #adb shell dumpsys package com.qianyilc.platform | grep -n 'version\|firstInstallTime'
    #停止应用
    #adb shell am force-stop package

#输出帮助信息
def printHelp():
    print('方法列表:')
    print('info             ·········  获取设备信息')
    print('apkinfo          ·········  获取指定包名的apk信息 支持模糊查询')
    print('ip               ·········  获取ip信息')
    print('text             ·········  将内容输入到edittext')
    print('packages         ·········  获取安装包列表 可跟关键字. eg:packages weibo')
    print('cap              ·········  截取当前屏幕,保存在当前目录,然后删除手机中的截图文件')
    print('screenrecord     ·········  录制当前屏幕,保存在当前目录,然后删除手机中的录屏文件')
    print('v 0.0.2 2017年03月06日16:55:36')

def test():
    result = _command('adb shell dumpsys package com.qianyilc.platform')
    print(result)
    match = re.findall('MAIN:.*:',result,flags=re.DOTALL+re.MULTILINE)
    print(match)
'''
    print (time.strftime("%Y_%m_%d_%H_%M_%S", time.localtime()))
    os.system(ads + 'input keyevent 64')
    os.system(ads + 'input keyevent 99')
    os.system(ads + 'input text aaaa')
'''

if len(sys.argv) ==1:
    deviceInfo()
elif sys.argv[1]=='info':
    deviceInfo()
elif sys.argv[1]=='cap':
    screencap()
elif sys.argv[1]=='ip':
    deviceIP()
elif sys.argv[1]=='packages':
    if len(sys.argv) ==2:
        packages('')
    else:
        packages(sys.argv[2])
elif sys.argv[1]=='help':
    printHelp()
elif sys.argv[1]=='screenrecord':
    screenrecord()
elif sys.argv[1]=='apkinfo':
    if len(sys.argv) ==2:
        apkInfo('')
    else:
        apkInfo(sys.argv[2])
elif sys.argv[1]=='text':
    if len(sys.argv) ==2:
        inputtext('')
    else:
        inputtext(sys.argv[2])
elif sys.argv[1]=='test':
    test()
else:
    print('未匹配到命令')
    printHelp()

File: adb_command.js
'use strict'
const fs = require('fs');
const path = require('path');
const program = require('commander');

const co = require('co')
const prompt = require('co-prompt')

const exec = require('child_process').exec
const spawn = require('child_process').spawn
const execSync = require('child_process').execSync

require('./base');
let YELLOW = '\x1b[33m'
let BLUE = '\x1b[34m'
let END = '\x1b[0m'

const AS = 'adb shell '

const run_as = function (command) {
    return run_command(AS + command);
}

const run_command = function (command) {
    let result = execSync(command);
    return result.toString().trim();
    //let rs = result.toString().replace('\r\n', '');
}

const run_commandA = function (command, callback) {
    let child = exec(command, function (error, stdout, stderr) {
        // console.log(error, stdout, stderr);
        if (!error) {
            if (callback)
                callback();
        }
    });
    child.on('exit', function (code) {
        if (code != 0) {
            console.log('Failed: ' + code);
        }
    });
}

const print_deviceinfo = function () {
    let brand = run_as('getprop ro.product.brand');
    let model = run_as('getprop ro.product.model');
    let name = run_as('getprop ro.product.name');
    console.log('品牌：' + run_as('getprop ro.product.brand'));
    console.log('型号：' + run_as('getprop ro.product.model'));
    console.log('代号：' + run_as('getprop ro.product.name'));
    console.log('系统版本：' + run_as('getprop ro.build.version.release') + ',' + run_as('getprop ro.build.version.sdk'));

    let pd = run_as('wm density');
    console.log('屏幕密度：' + pd.substring(pd.indexOf(':') + 1));

    let ps = run_as('wm size');
    console.log('屏幕分辨率：' + ps.substring(ps.indexOf(': ' + 1)));
}

const print_IP = function () {
    console.log(run_as('ifconfig | grep Mask'));
}

const select_packages = function (packagename, tip, callback) {
    if (packagename) {
        let rs = run_as(`pm list packages ${packagename} | awk -F ':' '{print $2}'`)
        let length = rs.split('\n').length;

        if (length === 0) {
            console.log(`抱歉,没有找个${packagename} 相关应用`);
        } else if (length === 1) {
            if (callback)
                callback(rs.split('\n')[0]);
        } else {
            rs.split('\n').forEach((item, index, array) => {
                console.log(index + ':' + item);
            });
            co(function *() {
                let line = yield prompt(tip)
                if (callback)
                    callback(rs.split('\n')[parseInt(line)]);
                process.stdin.pause();
            })
        }
        console.log();
    } else {
        console.log('抱歉 输入错误 -h 查看帮助');
    }
}

const print_packages = function (packagename) {
    select_packages(packagename, '请选择要查看的应用编号:', (name) => {
        doPpackageInfo(name);
    })
}

const doPpackageInfo = function (packagename) {
    console.log(packagename);
    //console.log(run_as(`dumpsys package ${packagename} `));
    console.log(run_as(`dumpsys package ${packagename} | grep -n 'version\\|firstInstallTime'`));
}

const cleanApk = function (packagename) {
    select_packages(packagename, '请选择要清除数据的应用编号:', (name) => {
        doCleanApk(name);
    })
}

const doCleanApk = function (packagename) {
    co(function *() {
        let conf = yield prompt.confirm('确定要清除<' + packagename + '>的数据?(y/n)')
        if (conf) {
            run_as('pm clear ' + packagename);
        }
        process.stdin.pause();
    })
}

const stopApk = function (packagename) {
    select_packages(packagename, '请选择要停止的应用编号:', (name) => {
        doStopApk(name);
    })
}

const doStopApk = function (packagename) {
    run_as('am force-stop ' + packagename);
}

const print_Log = function (tag) {
    let filename = 'log/' + new Date().format("yyyy-MM-dd-hh") + '.txt'
    fs.appendFile(filename, '##### 记录日志:' + new Date().format("yyyy-MM-dd hh:mm:ss") + '\n');

    let log = spawn('adb', ['logcat', `${tag}:v`, '*:S']);
    log.stdout.on('data', function (data) {
        console.log(data.toString());
        fs.appendFile(filename, data.toString());
    });
    // 添加一个end监听器来关闭文件流
    log.stdout.on('end', function (data) {
        console.log(' downloaded to ');
    });
}

const screencap = function () {
    let filename = new Date().format("yyyy_MM_dd_hh_mm_ss") + '.png';

    run_commandA(AS + 'screencap -p /sdcard/' + filename, () => {
        run_commandA('adb pull /sdcard/' + filename + ' record/', () => {
            run_commandA('adb shell rm /sdcard/' + filename, () => {
            });
        });
    });
}

const screenrecord = function () {
    console.log(process.platform);

    console.log('屏幕录制中...需要停止时按ctrl+c');
    let filename = new Date().format("yyyy_MM_dd_hh_mm_ss") + '.mp4';
    let log = spawn('adb', ['shell', 'screenrecord', '/sdcard/' + filename]);
    co(function *() {
        let conf = yield prompt.confirm('停止录制?(y/n)')
        if (conf) {
            log.kill();
            setTimeout(() => {
                run_commandA('adb pull /sdcard/' + filename + ' record/', () => {
                    run_commandA('adb shell rm /sdcard/' + filename, () => {
                    });
                });
            }, 1000);
        }
        process.stdin.pause();
    })
}

const setInput = function (content) {
    run_as('input text ' + content);
}

const _mkdir = function (dir) {
    fs.exists(dir, function (exists) {
        !exists && fs.mkdir(dir);
    });
}

_mkdir('log');
_mkdir('record');


program
    .version(require('./package.json').version)
    .option('-i, --info', '显示设备信息', print_deviceinfo)
    .option('-l, --log <tag>', '打印log(log/) <tag>', print_Log)
    .option('-p, --package <package>', '查看apk信息 <包名>', print_packages)
    .option('--clean <package>', '双清(缓存/数据)应用 <包名>', cleanApk)
    .option('--stop <package>', '杀死应用应用 <包名>', stopApk)
    .option('--input <content>', '将内容输入到edittext(不支持中文)', setInput)
    .option('-c --cap', '截屏并保存的电脑(record/)', screencap)
    .option('-s --screenrecord', '录制并保存的电脑(record/)', screenrecord)
    .option('--ip', '显示设备ip地址', print_IP)
    .parse(process.argv);

if (program.rawArgs.length <= 2) {
    console.error(0);
    program.help();
}